### ISSUE TITLE
Messare Rewrite Facility-like system

### ISSUE THREAD

**Issue by aumetra** [kitsune-soc/kitsune#160]
Somewhat of a followup to #140/#159 (since this system could potentially deprecate them)

I would really like to have a system similar to Akkoma's "Message Rewrite Facility" system (https://docs.akkoma.dev/stable/configuration/mrf/).  

Somewhat related to #132 since we have to figure out what "scripting" system we want for these facilities.  
The current main candidates are:

- WASM
- Lua

**Comment by erlend-sh** [comment]
WASM would be great (multi-lang support!), but the ecosystem is still in flux.

Another good candidate is Rhai: https://www.apollographql.com/docs/router/customizations/rhai

**Comment by silverpill** [comment]
+1 for WASM. I'm [considering](https://codeberg.org/silverpill/mitra/issues/38#issuecomment-928579) it as well, because WASM MRF modules can be reused by other implementations.

**Comment by aumetra** [comment]
@silverpill Interoperability would be really nice. I think an API such as 

$$
(retCode, newActivityOption) = f(activity)
$$

should be sufficient for a start? That should be possible since WASM does have the possibility to return multiple values now

**Comment by silverpill** [comment]
Yes, that should be enough. In the future it would be nice to also expose some kind of API to MRF module (to let it read some data from database, for example).

**Comment by silverpill** [comment]
That would make a good Fediverse Enhancement Proposal :)
I can't start working on it right away though, need to finish a bunch of other tasks first, and I haven't worked with WebAssembly before... So I will probably follow your lead.

**Comment by aumetra** [comment]
I guess that would also be a nice preparation for me to write an FEP before I attempt to write the "identity proof login" FEP, heh

**Comment by erlend-sh** [comment]
Related scripting discussions in Lemmy:
- https://github.com/LemmyNet/lemmy/issues/3562
- https://github.com/LemmyNet/lemmy/issues/3277
